doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,824 @@
+intermediate representation known as GOTO programs instead.
+
+\subsection Trace_section Trace
+
+TODO: I am not quite sure what is supposed to go here.
+
+From GitHub: "I think this was intended to describe how a (counterexample) trace is represented."
 
 \section analysis_techniques_section Analysis techniques
 
 \subsection BMC_section Bounded model checking
 
 To be documented (can copy from the CBMC manual).
 
+TODO: There seems to be no sufficiently comprehensive section in the
+CBMC manual. Find some other sources to collect information from.
+
+- Idea of model checking: Given some property, check if we can reach
+  it; if yes, under what conditions?
+- Bounded model-checking: Cut off search at some point; either prove
+  that property must occur before, or accept underapproximation.
+- How does it work? Encoding of state as propositions/first-order
+  propositions, step function and composition of step functions.
+- Solve question: Is there a model for this formula?
+
 \subsection SAT_section SAT and SMT
 
-To be documented.
+- Goal: Solve model-finding problem.
+- Answers: Here is a model, no model possible
+- Complexity issues
+- SAT and SMT as extremely advanced black boxes that solve
+  a good part of problems efficiently. Point out MiniSat,
+  Z3 as good examples that we use.
+- Pointer to literature: CDCL, Lazy/Eager instantiation
+
+From our wiki:
+In a Boolean formula, variables can only be assigned values from the set
+{true, false}, and they are connected using and, or and not operators.
+The Boolean satisfiability problem is to find a valuation of the variables
+of a given formula so that the formula evaluates to true, or to prove that
+there is no such valuation. SAT is an NP-complete problem, meaning
+a given solution can be always verified in an amount of time that is polynomial
+in terms of the size of the input formula, and
+there is no known algorithm that computes a solution to any given formula
+in polynomial time (and the existence of such an algorithm would prove P = NP).
 
 \subsection static_analysis_section Static analysis
 
-To be documented.
+While BMC analyses the program by transforming everything to logic
+formulas and, essentially, running the program on sets of concrete
+states, another approach to learn about a program is based on the idea
+of interpreting an abstract version of the program. This is known
+as **abstract interpretation**. Abstract interpretation is one of the
+main methods in the area of **static analysis**.
+
+The key idea is that instead of looking at concrete program states
+(e.g., ''variable x contains value 1''), we look at some
+sufficiently-precise abstraction (e.g., ''variable x is odd'', or
+''variable x is positive''), and perform interpretation of the program
+using such abstract values. Coming back to our running example, we wish
+to prove that the factorial function never returns 0.
+
+The first ingredient we need for abstract interpretation is the
```

Comment:
For clarity, I would list here all the ingredients you will need:
* An abstract domain, abstracting away details in the data
* A transformer per statement
* A map

FIXME